designs/node-overlay.md

```diff
@@ -0,0 +1,305 @@
+
+The fail closed approach, while providing strong guarantees and explicit failure signals, has significant operational drawbacks. By halting node launches until misconfigurations are resolved (either globally or per NodePool), it could unnecessarily impact cluster scalability and workload availability. This aggressive stance on configuration errors could turn minor overlay misconfigurations into broader service disruptions, particularly in scenarios where the misconfiguration affects only a subset of resources that aren't critical to immediate scaling needs. Since pods requiring specific overlay-defined resources would fail to schedule anyway (as they do today), the additional protection of failing closed provides minimal benefit while increasing operational risk.
+
+## Observability 
+
+Observability for Node Overlay will center around providing users with practical tools to understand the impact of their configurations. We will develop a diagnostic tool that allows users to preview how their Node Overlay configurations affect instance pricing and capacity allocations before deployment. This tool will help operators visualize and validate changes to instance selection patterns, price modifications, and capacity adjustments without requiring extensive logging or event generation. While we maintain basic metrics for overlay application status, this proactive tooling approach provides more actionable insights than high-volume logs or events, which could create operational overhead and scalability concerns in large clusters.
+
+```
+kind: NodeOverlay
+metadata:
+  name: price-modification
+spec:
+  requirements:
+  - key: karpenter.sh/capacity-type
+    operator: In
+    values: ["on-demand"]
+  priceAdjustment: 0.101 
+---
+kind: NodeOverlay
+metadata:
+  name: extended-resource
+spec:
+  requirements:
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ["c5.large"]
+  capacity:
+    smarter-devices/fuse: 1
+```
+
+```
+> karpenter-overlay diagnose --overlay price-modification extended-resource --nodepool default
+NODEPOOL    OVERLAY                      INSTANCE_TYPE   CAPACITY TYPE      PRICE    CAPCITY
+default     "price-modification"         m5.large        spot               0.104     
+default     "extended-resource"          c5.large        spot               0         smarter-devices/fuse: 1
+...
+```
+
+The integration of Node Overlay adjustment values and deeper observability mechanisms will be deferred until clear users requirements emerge. This decision is partly driven by concerns that excessive events or logs produced by some controllers could be highly noisy or negatively impact the API server. By deferring these features, we can focus on delivering core functionality and validation tooling first, while leaving room for future enhancements. These enhancements will be based on real-world usage patterns and specific users needs for additional visibility into Node Overlay operations.
+
+## Launch Scope
```

Comment:
I think you should also add something on the NodeClaim that calls out the overlay used

Reply:
> the hash of the overlay

What do you mean by the hash of the overlay? 

We're currently deferring the implementation of overlay status reporting until we receive more customer feedback to better understand the most valuable approach. While NodeClaim is one potential option, we're also considering exposing overlay information through NodePool status or the NodeOverlay status itself, including details about which instance types had overlays applied. We're particularly mindful of the signal-to-noise ratio, especially if we implement event or log-based tracking. What kind of visibility would be most useful for your use case?